flink-runtime/src/main/java/org/apache/flink/runtime/resourcemanager/slotmanager/SlotManagerImpl.java

```diff
@@ -230,30 +230,34 @@ public int getNumberFreeSlotsOf(InstanceID instanceId) {
 
     @Override
     public ResourceProfile getRegisteredResource() {
-        return getResourceFromNumSlots(getNumberRegisteredSlots());
+        return taskManagerRegistrations.values().stream()
+                .map(TaskManagerRegistration::getTotalResource)
+                .reduce(ResourceProfile.ZERO, ResourceProfile::merge);
     }
 
     @Override
     public ResourceProfile getRegisteredResourceOf(InstanceID instanceID) {
-        return getResourceFromNumSlots(getNumberRegisteredSlotsOf(instanceID));
+        return Optional.ofNullable(taskManagerRegistrations.get(instanceID))
+                .map(TaskManagerRegistration::getTotalResource)
+                .orElse(ResourceProfile.ZERO);
     }
 
     @Override
     public ResourceProfile getFreeResource() {
-        return getResourceFromNumSlots(getNumberFreeSlots());
+        return taskManagerRegistrations.keySet().stream()
+                .map(this::getFreeResourceOf)
+                .reduce(ResourceProfile.ZERO, ResourceProfile::merge);
```

Comment:
Better to iterate on the entry set than to iterate on the key set and get the entry with the key.